pilot/pkg/model/push_context.go

```diff
@@ -78,9 +80,127 @@ type PushContext struct {
 	// Env has a pointer to the shared environment used to create the snapshot.
 	Env *Environment `json:"-,omitempty"`
 
+	// ServiceAccounts is a function mapping service name to service accounts.
+	ServiceAccounts func(string) []string `json:"-"`
+
+	// ServicePorts is used to keep track of service name and port mapping.
+	// This is needed because ADS names use port numbers, while endpoints use
+	// port names. The key is the service name. If a service or port are not found,
+	// the endpoint needs to be re-evaluated later (eventual consistency)
+	ServiceName2Ports map[string]map[string]uint32 `json:"-"`
+	ServicePort2Name  map[string]map[uint32]string `json:"-"`
+
 	initDone bool
 }
 
+// XDSUpdater is used for direct updates of the xDS model and incremental push.
+// Pilot uses multiple registries - for example each K8S cluster is a registry instance,
+// as well as consul and future EDS or MCP sources. Each registry is responsible for
```

Comment:
Please add: "Each registry forms a single virtual network boundary for purposes of connectivity".

Reply:
Except it's not true :-)

A registry may support multiple networks.

A separate structure organized by network will be added for the 'split horizon' - this is just for optmizing incremental endpoint push.

Reply:
Show me one example of a registry (today, not in some distant future) that has multiple networks and requires gateways to talk to one another? The implication of your argument is that without istio, the endpoints in this registry cannot talk to one another at all, which is certainly not the case. 

I have thought about this long and hard.. The network boundary, today, essentially conforms to a cluster boundary. And the code is also doing the same. Adding yet another network boundary creates immense confusion for no good.

Reply:
Grouping by network is not only for the purpose of zvpn/gateway. Locality is another criteria - and I believe GKE and bluemix have support for k8s clusters spanning multiple zones. So does consul. 

So even if you have a single, flat network - the endpoints don't need to be all be in the same list.

The 'split horizon' is intended not only for the gateway traversal, but also for returning EDS results by locality. 

Also this PR, once again, doesn't introduce any network boundary. It just breaks up the big list of 
endpoints into shards, so they can be updated and manipulated with less overhead.

Reply:
>Grouping by network is not only for the purpose of zvpn/gateway. Locality is another criteria - and I believe GKE and bluemix have support for k8s clusters spanning multiple zones. So does consul. So even if you have a single, flat network - the endpoints don't need to be all be in the same list.

You can either have all endpoints directly talking to one another or talk to each other via a gateway. So, whether or not you group endpoints by locality, if you are going to be sending all endpoints to an Envoy (when one endpoint in a region/network changes), then how does the grouping help? Its not like we are sending just the endpoint that has changed. Even in this PR, you are sending all endpoints to envoys aren't you?